### PR TITLE
[FLINK-32549][network] Tiered storage memory manager supports ownership transfer for buffers

### PR DESCRIPTION
**Pull request by TanYuxin-tyx**
## What is the purpose of the change

Currently, the accumulator is responsible for requesting all buffers, leading to an inaccurate number of requested buffers for each tier.
To address this issue, buffer ownership must be transferred from the accumulator to the tiers when writing them, which will enable the memory manager to maintain a correct number of requested buffers for different owners.

## Brief change log

  - *Tiered storage memory manager supports ownership transfer for buffers*
 
**Version 1 (Deprecated. Do not use this way to fix anymore)**:
1. In order to transfer ownership, added a new API `TieredStorageMemoryManager#transferBufferOwnership`.
2. If `TierProducerAgent#tryWrite` returns `true` (indicating the write is successful), then 
 the buffer has been written to the tier(once written, the buffer may be consumed or flushed immediately, so the buffer recycler can not be updated after `TierProducerAgent#tryWrite`), `TieredStorageProducerClientImpl#writeAccumulatedBuffer` can not update the buffer recycler anymore, so the updating buffer recycler process is moved into each tier.
3. Because the updating buffer recycler process is moved into each tier and the buffer recycler can not be updated after compression, the buffer compress process is also moved into each tier.
4. To use the right buffer recycler, added a new API `TieredStorageMemoryManager#getBufferRecycler`.

**Version 2**:
1. `Buffer` added a new API `setRecycler`
2. Added a new API for memory manager `TieredStorageMemoryManager#transferBufferOwnership`

## Verifying this change

This change added tests and can be verified as follows:

  - *Added tests to verify the buffer ownership transfer in `TieredStorageMemoryManagerImplTest`*
- *Added tests to verify getting the buffer recycler in `TieredStorageMemoryManagerImplTest`*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

**Comment by xintongsong** [comment]
@TanYuxin-tyx, I have 2 questions regarding the current approach.
1. The transfer of the buffer ownership seems to be divided into two steps: a) updating the buffer recycler, and b) updating the bookkeepings in memory manager. Currently, the two steps take place in different positions. Why is that necessary?
2. Why is updating of recycler not possible after the buffer being compressed?

**Comment by TanYuxin-tyx** [comment]
@flinkbot run azure

**Comment by TanYuxin-tyx** [comment]
@xintongsong Thanks for helping review. Based on the offline discussions, I have updated the change. In the new PR, the above issues no longer exist anymore. PTAL, thanks.

The change details are as follows.
1. Buffer added a new API `setRecycler`.
2. Added a new API for memory manager `TieredStorageMemoryManager#transferBufferOwnership`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TieredStorageMemoryManagerImpl.java [hunk 1]
```diff
@@ -276,6 +294,13 @@ private void recycleBuffer(Object owner, MemorySegment buffer) {
         decNumRequestedBuffer(owner);
     }
 
+    private BufferRecycler getOwnerBufferRecycler(Object owner) {
+        return memorySegment -> {
+            bufferPool.recycle(memorySegment);
+            decNumRequestedBuffer(owner);
+        };
+    }
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TieredStorageMemoryManagerImpl.java:302]: This seems to be duplicated with `recycleBuffer`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/disk/DiskCacheManager.java [hunk 2]
```diff
@@ -41,6 +41,8 @@ class DiskCacheManager {
 
     private final int numSubpartitions;
 
+    private final TieredStorageMemoryManager memoryManager;
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/disk/DiskCacheManager.java:44]: What is the purpose of this change? The filed seems never accessed after initialized.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TieredStorageMemoryManagerImpl.java [hunk 3]
```diff
@@ -200,6 +202,22 @@ public int numOwnerRequestedBuffer(Object owner) {
         return numRequestedBuffer == null ? 0 : numRequestedBuffer.get();
     }
 
+    @Override
+    public void transferBufferOwnership(Object oldOwner, Object newOwner, Buffer buffer) {
+        if (!buffer.isBuffer()) {
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (TanYuxin-tyx) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TieredStorageMemoryManagerImpl.java]: Moved the filter of `is null` outside the method.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/disk/DiskCacheManager.java (hunk 2 above)
Comment (TanYuxin-tyx) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/disk/DiskCacheManager.java]: Fixed.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TieredStorageMemoryManagerImpl.java [hunk 4]
```diff
@@ -200,6 +201,20 @@ public int numOwnerRequestedBuffer(Object owner) {
         return numRequestedBuffer == null ? 0 : numRequestedBuffer.get();
     }
 
+    @Override
+    public void transferBufferOwnership(Object oldOwner, Object newOwner, Buffer buffer) {
+        checkState(buffer.isBuffer(), "Only buffer supports transfer ownership.");
+        AtomicInteger numOldOwnerRequested = numOwnerRequestedBuffers.get(oldOwner);
+        if (numOldOwnerRequested == null) {
+            throw new RuntimeException("Failed to transfer buffer ownership for " + oldOwner);
+        }
```
Comment (reswqa) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TieredStorageMemoryManagerImpl.java:210]: This can be replaced by `checkNotNull`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/Buffer.java [hunk 5]
```diff
@@ -87,6 +87,13 @@ public interface Buffer {
      */
     BufferRecycler getRecycler();
 
+    /**
+     * Sets the buffer's recycler.
+     *
+     * @param bufferRecycler the new buffer recycler
+     */
+    void setRecycler(BufferRecycler bufferRecycler);
```
Comment (reswqa) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/Buffer.java:95]: This is an unsafe operation, we may need add more descriptions of this method to avoid others accidentally introducing bugs. Especially for slice buffer.

**Review by reswqa — COMMENTED** [review]
Thanks for the fix, this changes overall looks good to me. I only left some minor comments.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TieredStorageMemoryManagerImpl.java (hunk 4 above)
Comment (TanYuxin-tyx) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TieredStorageMemoryManagerImpl.java]: Use `decNumRequestedBuffer(oldOwner)` instead.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/Buffer.java (hunk 5 above)
Comment (TanYuxin-tyx) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/Buffer.java:100]: Fixed. Added docs to describe the method.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TieredStorageMemoryManagerImpl.java [hunk 6]
```diff
@@ -196,8 +197,16 @@ public int getMaxNonReclaimableBuffers(Object owner) {
 
     @Override
     public int numOwnerRequestedBuffer(Object owner) {
-        AtomicInteger numRequestedBuffer = numOwnerRequestedBuffers.get(owner);
-        return numRequestedBuffer == null ? 0 : numRequestedBuffer.get();
+        Integer numRequestedBuffer = numOwnerRequestedBuffers.get(owner);
+        return numRequestedBuffer == null ? 0 : numRequestedBuffer;
```
Comment (reswqa) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TieredStorageMemoryManagerImpl.java:201]: Can `numOwnerRequestedBuffers.getOrDefault(owner, 0)` simplify this?

**Review by reswqa — APPROVED** [review]
Thanks for the update, LGTM.
